src/main/java/com/google/step/datamanager/DealManagerDatastore.java

```diff
@@ -146,37 +149,56 @@ public Deal updateDeal(Deal deal, List<String> tagNames) {
     return transformEntityToDeal(dealEntity);
   }
 
-  /** Retrieves deals posted by restaurants or users */
-  private List<Deal> getDealsPublishedByRestaurantsOrUsers(
-      Set<Long> ids, String filterAttribute, int limit) {
-    List<Deal> dealResults = new ArrayList<>();
+  /** Retrieves deals with a filter applied */
+  private List<Long> getDealsWithFilter(
+      List<Object> ids, String filterAttribute, int limit, String sort) {
+    List<Long> dealResults = new ArrayList<>();
     if (ids.size() > 0) {
       Filter propertyFilter = new FilterPredicate(filterAttribute, FilterOperator.IN, ids);
-      Query query = new Query("Deal").setFilter(propertyFilter);
+      Query query = null;
+      Iterable<Entity> entities = null;
+      if (sort != null) {
+        query =
+            new Query("Deal")
+                .setFilter(propertyFilter)
+                .addSort("timestamp", SortDirection.DESCENDING);
+      } else {
+        query = new Query("Deal").setFilter(propertyFilter);
+      }
       PreparedQuery pq = datastore.prepare(query);
-      Iterable<Entity> entities;
       if (limit > 0) {
         entities = pq.asIterable(FetchOptions.Builder.withLimit(limit));
       } else {
         entities = pq.asIterable();
       }
       for (Entity entity : entities) {
-        dealResults.add(readDeal(entity.getKey().getId()));
+        dealResults.add(entity.getKey().getId());
       }
     }
     return dealResults;
   }
 
   /** Retrieves deals posted by users */
   @Override
-  public List<Deal> getDealsPublishedByUsers(Set<Long> userIds, int limit) {
-    return getDealsPublishedByRestaurantsOrUsers(userIds, "posterId", limit);
+  public List<Long> getDealsPublishedByUsers(Set<Long> userIds, int limit, String sort) {
+    return getDealsWithFilter(new ArrayList<>(userIds), "posterId", limit, sort);
+    // List<Long> list = new ArrayList<>(Arrays.asList(1L, 2L));
+    // List<Object> list1 = list;
   }
 
   /** Retrieves deals posted by restaurants */
   @Override
-  public List<Deal> getDealsPublishedByRestaurants(Set<Long> restaurantIds, int limit) {
-    return getDealsPublishedByRestaurantsOrUsers(restaurantIds, "restaurantId", limit);
+  public List<Long> getDealsPublishedByRestaurants(
+      Set<Long> restaurantIds, int limit, String sort) {
+    return getDealsWithFilter(new ArrayList<>(restaurantIds), "restaurantId", limit, sort);
+  }
+
+  /** Retrieves deals given a set of ids */
+  @Override
+  public List<Long> getDealsWithIds(Set<Long> ids, int limit, String sort) {
+    List<Key> keys =
+        ids.stream().map(id -> KeyFactory.createKey("Deal", id)).collect(Collectors.toList());
+    return getDealsWithFilter(new ArrayList<>(keys), "__key__", limit, sort);
```

Comment:
no need to write new ArrayList<>() since it's List<> already?

Reply:
I need to convert to array list since if not i wouldnt be able to cast the List<Key> to the List<Object>

Reply:
oh i see